primitives/src/xvm.rs

```diff
@@ -33,56 +33,100 @@ pub enum VmId {
 }
 
 impl TryFrom<u8> for VmId {
-    type Error = CallError;
+    type Error = FailureReason;
 
     fn try_from(value: u8) -> Result<Self, Self::Error> {
         if value == VmId::Evm as u8 {
             Ok(VmId::Evm)
         } else if value == VmId::Wasm as u8 {
             Ok(VmId::Wasm)
         } else {
-            Err(CallError::InvalidVmId)
+            Err(FailureReason::Error(FailureError::InvalidVmId))
         }
     }
 }
 
 /// XVM call info on success.
 #[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
-pub struct CallInfo {
+pub struct CallOutput {
     /// Output of the call.
     pub output: Vec<u8>,
     /// Actual used weight.
     pub used_weight: Weight,
 }
 
-/// XVM call error on failure.
+impl CallOutput {
+    /// Create a new `CallOutput`.
+    pub fn new(output: Vec<u8>, used_weight: Weight) -> Self {
+        Self {
+            output,
+            used_weight,
+        }
+    }
+}
+
+/// XVM call failure.
 #[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
-pub enum CallError {
-    /// Invalid VM id.
-    InvalidVmId,
-    /// Calling the contracts in the same VM is not allowed.
-    SameVmCallDenied,
+pub struct CallFailure {
+    /// Failure reason.
+    pub reason: FailureReason,
+    /// Actual used weight.
+    pub used_weight: Weight,
+}
+
+impl CallFailure {
+    /// Create a new `CallFailure` on revert.
+    pub fn revert(details: FailureRevert, used_weight: Weight) -> Self {
+        Self {
+            reason: FailureReason::Revert(details),
+            used_weight,
+        }
+    }
+
+    /// Create a new `CallFailure` on error.
+    pub fn error(details: FailureError, used_weight: Weight) -> Self {
+        Self {
+            reason: FailureReason::Error(details),
+            used_weight,
+        }
+    }
+}
+
+/// Failure reason of XVM calls.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum FailureReason {
+    /// XVM call failed with explicit revert.
+    Revert(FailureRevert),
+    /// XVM call failed with error.
+    Error(FailureError),
+}
+
+/// Failure reason on revert.
```

Comment:
In the context of XVM, what is the difference between `Error` and `Revert`?

Reply:
`Error` is the case that execution failed and must stop. `Revert` is the execution succeeded but the callee contract explicitly asks to revert. I also added input errors like `InvalidTarget` to `Revert`, as from EVM/WASM perspective it should be a defensive guard condition instead of an execution failure.

 `Error` maps to EVM `ExitError/ExistFatal` and WASM `DispatchError`. `Revert` maps to EVM `ExitRevert` and WASM `REVERT` flag.

Reply:
Docstring added. I also pushed a commit with contract address variables renaming in tests, to make it more readable.